beater/v2_handler.go

```diff
@@ -108,8 +116,22 @@ func (v *v2Handler) Handle(beaterConfig *Config, report publish.Reporter) http.H
 			v.sendResponse(logger, w, &sr)
 			return
 		}
+		ctx := r.Context()
+
+		// fetch the rate limiter from the cache, if a cache is given
+		var limiter *rate.Limiter
+		if v.cache != nil && beaterConfig.RumConfig.EventRate != nil {
+			key := utility.RemoteAddr(r)
+			if !v.cache.Contains(key) {
```

Comment:
Looks like there's a [TOCTOU](https://en.wikipedia.org/wiki/Time_of_check_to_time_of_use) bug here. It's technically possible, if unlikely, for the entry to be evicted after the `if` block but before the subsequent `cache.Get`, which would then return nil.

I'd suggest this instead:

```go
limiter, ok := v.cache.Get(key)
if !ok {
    limiter = rate.NewLimiter(...)
    v.cache.Add(key, limiter)
}
ctx = context.WithValue(r.Context(), stream.RateLimiterKey, limiter)
```

Reply:
Also, looks like there's a attack possible here, due to use of LRU. Say a DDoS is taking place, then the LRU is filled up with many IPs. new IPs coming in will be given a new rate limiter with a full capacity/burst.

I think what you want to do is check if an item was evicted, and reuse that rate limiter, so you'll probably get an empty bucket.

Reply:
Ah. Nice catch

Reply:
@axw I agree that the usage of an LRU cache introduces a possible attack surface. However, I brought this up in discussions with @roncohen earlier this week, where he argued for keeping the LRU cache. The alternative would be to use a cache with expiration date, that is either unbounded (could lead to endless memory usage) or limited to a certain number of entries, in which case reaching the number of entries would also be problematic. 

Regarding your proposed solution to avoid the TOCTOU, I agree that your suggestion gets rid of this problem. Both of our solutions could potentially lead to race conditions though. I'll think about this some more.

Reply:
I think there's a couple of different discussions going on here. There's the question about when to call `Wait`/`WaitN` and then there's the question of whether we should get the rate limiter from the LRU when we want to call `Wait`/`WaitN` or we should get the rate limiter from the LRU when the request comes in. 

I [argued](https://github.com/elastic/apm-server/pull/1367#issuecomment-420216773) that if there's a significant performance impact from calling `Wait` for every event, a potential way to solve it would be to consult the rate limiter for each batch instead. If there's no significant performance impact, i think it's much simpler to consult the rate limiter for each event.

I asked whether it would be [simpler](https://github.com/elastic/apm-server/pull/1367#discussion_r216604275) to get the rate limiter from the LRU and store it in the context instead of storing the LRU key in the context. If it's not simpler, I'm fine with getting the rate limiter from the LRU every time.

Reply:
@axw could you elaborate on this?: 
> I think what you want to do is check if an item was evicted, and reuse that rate limiter, so you'll probably get an empty bucket.

I believe @axw suggested offline if we set the LRU size to the same size as the overall `concurrent_requests` limit that we'll implement for v2, that would mitigate the attack.

Reply:
> I believe @axw suggested offline if we set the LRU size to the same size as the overall concurrent_requests limit that we'll implement for v2, that would mitigate the attack.

That just ensures that the bucket for the current IP won't get evicted mid-way through a request while retaining an upper bound on the number of active rate limiter objects.

>> I think what you want to do is check if an item was evicted, and reuse that rate limiter, so you'll probably get an empty bucket.

Sorry, a bit cryptic. And this is all a bit hypothetical, I haven't actually implemented this before.

Say your LRU has size N, and an attacker has 2*N IPs at their disposal. They cycle through the first N IPs until they get rate-limited, and then cycle through the next N, and then switch back to the first N and so on.

The reason that's an issue is because after N IPs have entered the LRU, then the N+1'th IP will evict one, and as it is implemented, a fresh limiter created with a full bucket, so you can burst immediately. What you could do instead is effectively have a pool of limiters by storing the evicted limiter using lru.NewWithEvict, and then reusing it when lru.Add returns true indicating another entry was evicted.

In pseudo-code:

```
var limiterPool sync.Pool
onEvicted := func(key interface{}, value interface{}) {
    limiterPool.Put(*value.(**rate.Limiter))
}
cache, err := lru.NewWithEvict(N, onEvicted)
if err {...}
...
var limiter *rate.Limiter
value, ok := cache.Get(key)
if ok {
    limiter = *value.(**rate.Limiter)
} else {
    if _, evicted := cache.Add(key, &limiter); evicted {
        limiter = limiterPool.Get().(*rate.Limiter)
    } else {
        limiter = rate.NewLimiter()
    }
}
```

Reply:
@roncohen thanks for clarification!

The two of us had an offline discussion, after your PR comments, where we discussed a couple of the comments and these were my take aways:
(1) use `batchsize` for checks if it performs better than checking for every request.
(2) fetch limiter once from cache and not for every event, because you didn't see an advantage for fetching it for every event but found the code to be more complicated. 
(3) don't use another form of cache, but stick with LRU cache. Make the limit configurable and document the downside in the config file. 
(4) get rid of checking whether or not the request itself should be allowed, as an additional limit and check will be applied for that. 

add (1): I did performance checks and `Wait` performs poorly so I switched to check in batches. However, also doing performance checks for `Reserve` I am going to switch back, as this behaves better.

add (2):  I might have misunderstood you there, as I thought you definitely prefer to not fetch the limiter per event but per request. We discussed how a request with many events get more hits in the LRU cache than a request with less events if the limiter is fetched for every event. So this is not only abut code simplicity but about a different semantic.

add (3): The suggestion to align the cache key limit with the amount of concurrent requests helps to ensure a key doesn't get evicted while fetching. It cannot solve the problem that an attack could bypass the whole rate limiting if requests are sent from multiple different ips, where `number_ips > cache_keys`.

ad (4): Introducing an additional `requests_per_ip_per_second` is fine with me, although I tried to avoid this additional config option.

Reply:
I just saw @axw 's last comment. @roncohen that is exactly the attack vector that we discussed offline. Since it is an LRU cache you can bypass with requests from too many different IPs. We also discussed this could be especially a problem with RUM agents on high traffic websites. 
I'll look into @axw 's suggestion with the callback.

Reply:
@axw I looked into your suggestion to reuse evicted limiters and I am not sure this is the way to go. My main issue with it is that this could lead to a lot of slow down for requests from IP addresses that actually do respect the rate limiting, while potentially allowing a larger amount of events from other IPs than intended. 

I also see an issue with re-using an old rate limiter for one request, as the next request for the same IP will still use a new rate limiter

Reply:
> I also see an issue with re-using an old rate limiter for one request, as the next request for the same IP will still use a new rate limiter

That sounds odd, can I see a diff somewhere?

Reply:
I was referring to your pseudo code: 
```
var limiter *rate.Limiter
if _, evicted := cache.Add(key, &limiter); evicted {
        limiter = limiterPool.Get().(*rate.Limiter)
    }
```
The limiter fetched from the `limiterPool` is not added to the pool again. So for the next request another limiter is used. But as cache keys do not expire here, it would be another evicted limiter and not a new one.